birdhouse/config/thredds/service-config.json.template

```diff
@@ -1,7 +1,7 @@
 {
     "name": "thredds",
     "url": "https://${PAVICS_FQDN_PUBLIC}/thredds",
-    "type": ["thredds"],
+    "type": ["data"],
```

Comment:
No `service-thredds`? Also, missing `catalog`? Based on the terms of the pattern in test schema.

Reply:
> Also, missing catalog? 

That sort of depends on the intended use-case for a lot of these. Once stac is in place then we intend to use that as the user-facing catalog and thredds will just be used to serve data.

In fact a lot of these will depend on the specifics of the permissions in magpie. For example: geoserver can optionally have wms, wfs, wps depending on the permission settings enabled. 

So the question ends up being: 

- does type indicate all _possible_ use cases?
- does type indicate all _intended_ use cases?
- does type indicate all _enabled_ use cases? (this gets complicated as permissions for users may differ)

I'm interested in any opinions on this. But I'm also just happy to keep things simple for now and go with all _possible_ use cases if that simplifies things for now.

Reply:
After a quick discussion with Alex who's writing code to consume this information, we think that indicating all _possible_ use cases is sufficient and easiest to describe and maintain.

I'll go for that for now.

Reply:
I don't think there is a big distinction between *intended*/*enabled* (as in the intention will drive what gets enabled or not), but I think this information is more relevant than *possible* use cases from a user points of view that will interact with the node. Basically, a user is interested in knowning what they can achieve with the node. I think specifying that a service *could* do something, but doesn't support it, is somewhat pointless for the users or developers interacting with that node. Listing all *possible* cases also somewhat poses a burden on the node provider, as in, it would be expected that they support all use cases, which is not something that is required nor that we expect.

Reply:
While I agree in theory, I don't see how we can implement this in practice.

The difference between intended and enabled are:

- intended: this service is available for some users of the node
- enabled: this service is available to *you*, a specific user of the node

One option would be to tie this in to the magpie providers definitions for this node but right now not all services are necessarily using magpie/twitcher for authentication.

If we wait until all services are authenticated through magpie we could take that strategy. In the meantime though, we need some solution to move this PR forward.

- if we're comfortable with keeping the "possible" strategy for now: this PR is ready
- if we'd prefer the "intended" strategy: we need to wait until all service are protected by magpie/twitcher so I suggest we remove the "type" key for now and implement it later
- if we'd prefer the "enabled" strategy: we need to go back to the drawing board and rethink the entire mechanism that this endpoint uses to provide the information about each service so I suggest we remove the "type" key for now and implement it later

Reply:
> I don't see how we can implement this in practice.

Aren't new nodes expected to be added by providers that want to register it? So they would be the ones to open a PR on the node-registry to add their instance. Only them should know exactly what they support. This could also include addionnal services and functionalities that are not part of the existing set in birhdouse-deploy.

I'm still failing to see any use case where a user browsing for node capabilities would even care about *possible* features. Only what they can achieve right away on a given node matters.

Reply:
You're talking about the "intended" strategy here, and you're right about the details of how this is implemented. 

I'm saying that we can't implement the "enabled" strategy since users will not be logged in when searching for services.

Even with the "intended" strategy, this gets more complicated. For example, consider the following scenario:

- a provider registers geoserver in magpie with the following configuration:

   ```yaml
   providers:
      geoserver:
        url: http://localhost:1234
        type: geoserver
        configuration:
          wfs: true
          wms: false
          wps: false
   ```

- we'd expect that the "type" key would contain `["services-wfs", "data"]` when the static json string is created when the proxy container starts up.
- if later the provider changes it to `wms: true` then we'd expect the "type" key to change to include `"service-wms"` but this will only be picked up by changing the json template file and restarting the proxy container.
- we could make it more automatic by generating the "types" from the magpie configuration

Reply:
That being said, I don't care that much about how this is implemented. We can adapt the front end to work with whatever data is provided by the services endpoint. 

@fmigneault please let me know what you would like to see here and we'll go with that so that we can move forward with this PR.

Reply:
Maybe were are confusing the terms between each other. I don't really mind about the "name" of the strategy. 
The only thing important is that it should reflect what a user can achieve when interacting with the node.

I don't think relying on Magpie configuration is the right way to go.
In the case of the birdhouse-deploy configuration provided in this repository, we can parse which WFS/WMS/WPS or any other provider are activated. However, another DACCS node with overrides will not necessarily expose those custom configurations.

Therefore, I can see only 2 valid options about what we can do to report node capabilities:
1. Rely on the node managers to open a PR on the node-registry to provide updates to the reference repository about the modified capabilities of their services for their node.
2. Have some kind of capabilities endpoint, similar to `/components` that provides directly the contents what the node-registry is expecting (schema to be completed/validated). The `/conformance` endpoint is quote common now of OGC and STAC APIs. Probably a similar approach with normative URI references in a `conformsTo` list could be defined. 

Either way, if they don't provide the capabilities updates which either approach, it is sadly "just too bad" for their node.

Reply:
> 1. Rely on the node managers to open a PR on the node-registry to provide updates to the reference repository about the modified capabilities of their services for their node.

This is a fine fallback if needed but I can see there being a lot of issues due to human error that I'd like to avoid if possible. 

>  2. Have some kind of capabilities endpoint, similar to /components that provides directly the contents what the node-registry is expecting (schema to be completed/validated).

This is what the `/services` endpoint is supposed to be. 

>  The /conformance endpoint is quote common now of OGC and STAC APIs. Probably a similar approach with normative URI references in a conformsTo list could be defined.

 I don't really understand how a conformance endpoint helps us achieve what we want here. I checked out the example in weaver (https://pavics.ouranos.ca/weaver/conformance) and environment canada (https://api.weather.gc.ca/conformance) and it is not clear to me how that is a solution to what we are trying to achieve here. Can you help me understand how adding a conformance endpoint helps us achieve our goal:

- We are trying to logically organize services into categories so that we can group them together for documentation purposes in a way that makes sense to the end user (paraphrased from here: https://github.com/DACCS-Climate/DACCS-node-registry/pull/7#issuecomment-1578693022)

One last thing I will say: being able to categorize services in this way is _not_ a requirement for the project. If we think that this sort of categorization is misleading since we can't define the categories well enough then we should _not_ include them at all. The json object would then only include "name", "url", "description", "documentation" as keys which would be entirely sufficient.

Reply:
> I don't really understand how a conformance endpoint helps us achieve what we want here.

If using the birdhouse-deploy stack as is, it is possible to know "somewhat" which services types/categories apply. 
However, this is technically not guaranteed to be correct or sufficient information to use the services, and it is not really that clear which subset of capabilities are supported.

For example, I could toggle off the `wms` capabilities of my instance's `geoserver`, and then, default services types/categories returned by `geoserver` on `/services` would incorrectly report that my node supports `service-wms`, when it doesn't.
The https://github.com/bird-house/birdhouse-deploy/blob/service-version-endpoints/birdhouse/config/geoserver/service-config.json.template file therefore needs to be dynamically populated, which becomes quickly very complicated. However, this is not even the worst case.

The goal of the `/conformance` endpoint is to be more explicit about the exact capabilities supported by each service.
The one I know the most is for the case of `weaver`. The *OGC API - Processes* is technically sub-divided into:
- https://github.com/opengeospatial/ogcapi-processes/tree/master/core (Part 1: Core)
- https://github.com/opengeospatial/ogcapi-processes/tree/master/extensions/deploy_replace_undeploy (Part 2: DRU)
- https://github.com/opengeospatial/ogcapi-processes/tree/master/extensions/workflows (Part 3: Workflows)
- https://github.com/opengeospatial/ogcapi-processes/tree/master/extensions/quotation
  https://github.com/opengeospatial/ogcapi-processes/tree/master/extensions/billing (WIP: Quotation & Billing)

And each of those parts are themselves sub-divided in a lot of conformances options that can be optionally supported.
There are similar sets of conformance URLs for WFS, STAC, etc. Some of them even overlap, since STAC is a superset around *OGC API - Features*, which is the predecessor of WFS (and partially compatible with it). 

In summary, I think having the general categorization is a good idea. We should keep that. 
However, I wonder if we could take advantage of the widely used approach of `/conformance` instead of redefining our own structure, to basically list similar "capabilities" (although less verbose for the moment). If this `/conformance` was used, many services would not need to update anything as they already provide it, as you illustrated with Weaver and Env Canada. STAC API also provides this list, and its a requirement for all new *OGC API*s.

Reply:
I have a better idea. Let's keep `/services`, but also allow a `links` section.
For https://github.com/bird-house/birdhouse-deploy/blob/service-version-endpoints/birdhouse/components/weaver/service-config.json.template, it would then be:
```json
{
  "name": "weaver",
  "type": [
    "service-ogcapi_processes"
  ],
  "description": "An OGC-API flavored Execution Management Service",
  "links": [
    { 
      "rel": "service",
      "type": "application/json",
      "href": "https://${PAVICS_FQDN_PUBLIC}/${WEAVER_MANAGER_NAME}/"
    },
    { 
      "rel": "service-desc",
      "type": "text/html",
      "href": "https://pavics-weaver.readthedocs.io/"
    },
    { 
      "rel": "service-doc",
      "type": "application/json",
      "href": "https://${PAVICS_FQDN_PUBLIC}/${WEAVER_MANAGER_NAME}/"
    },
    {
      "rel": "conformance",
      "type": "application/json",
      "href": "https://${PAVICS_FQDN_PUBLIC}/${WEAVER_MANAGER_NAME}/conformance/"
    }
  ]
}
```
This uses valid/recommended https://www.iana.org/assignments/link-relations/link-relations.xhtml names and is extensible.

Reply:
Given that the list of `conformance` items could be identified from some provided link, I would worry less about the specific values of `type`. They would almost be the same as `keywords`, which are common in other APIs (but with a more restrictive set to your linking). 

In the case of GeoServer and the birds, that would be each of the WFS, WMS, etc. `GetCapabilities` endpoint that could return what they support (see links in https://pavics.ouranos.ca/geoserver/web/). If OGC APIs were added to GeoServer, they would each report their own `/conformance` endpoint as well (https://geoserver.geosolutionsgroup.com/edu/en/ogcapi/intro.html#openapi-based).

In THREDDS, the best I could find was: https://pavics.ouranos.ca/testthredds/catalog.xml
So at least, we have some form of listing of reference schemas and capapilities it supports.

I don't expect "conformance" to be a *requirement* for all services, just that it is available whenever possible.

Reply:
@fmigneault The point of this PR is simply to expose information that will be consumed by the node-registry, and some DACCS components that use the registry, such as the front-end. There are, so far, no other meaningful uses of the `/services` endpoint by anyone else, for anything else. Given this limited use case, it seems unnecessary to get into the weeds of what something should be called and the conformance standards that should be used. We could all simply agree to go with calling things whatever we want for now and revise things later.

In fact, I do not see why we cannot simply go with @mishaschwartz's proposal to expose all possible capabilities at this point and then in future, should the need arise for one reason or another, constrain the `types` field to intended or enabled capabilities. I see this as a more agile way of doing things and letting experience and use cases dictate how the behaviour of the endpoint evolves, rather than beating the issue to death in this PR before we have even gotten started to using the feature. 

I think the `links` format makes sense, and I think an amendment to this PR that includes that change should be sufficient to warrant merging this PR and moving forward.

Reply:
@dchandan 

> There are, so far, no other meaningful uses of the `/services` endpoint by anyone else, for anything else. 

But should we limit ourselves to this narrow mindset because of this?
While designing a registry to centralize the node definitions, we should try to identify ahead of time possible issues, and emphasise on approaches that will favor reusing this information effectively. Otherwise, we end up in a similar situation as ESGF, were we have to patch up a metagrid definition after the fact to address the huge mess of capabilities from each node.

>  I do not see why we cannot [...] expose all possible capabilities at this point

This is exactly what I'm trying to achieve. I'm simply pointing out that saying for example `service-ogcapi_processes` is not sufficient to indicate what this service is capable of. It is much more complicated than that. 

I'm simply sharing experience I have gathered over time from working with OGC and other Geo-related standards. 
Instead of reinventing the wheel and making a custom schema, I propose alternative methods to represent the same data, but that just so happens to align with existing web best pratices used by many APIs.

If the goal is to allow nodes to interact with each other and be able to properly communicate their capabilities, they should be able to provide this information is the most descriptive manner possible. If the way to represent these references always changes, this will become a huge blocker for interroperability.

Reply:
@fmigneault Sure that's fine, we'll use the links key as you suggest. 

Based on your comments I'm going to assume that your answer to the question I raised in this comment (https://github.com/bird-house/birdhouse-deploy/pull/329#discussion_r1221550632):

> Are you saying that you don't care what is in the type key if we change the links key to your suggestion?
If so, please let me know and I'll update the links key to your suggestion and I'll decide what goes in the type key.

is: "use the format described here (https://github.com/bird-house/birdhouse-deploy/pull/329#discussion_r1220478936) and leave the current values of the `type` key as is". 

If that is not what you meant, please let me know as soon as possible so that I can proceed to update this PR.

Reply:
I would like to have the `ilnks` provided for reference, but you are free to use the `type` as well with whichever set of values that fills your needs for grouping the services in the node-registry.